File: src/main/scala/kim/dencka/easy/linked_list/MiddleOfTheLinkedList.scala
package kim.dencka.easy.linked_list

import LINKED_LIST_STRUCTURE._

object MiddleOfTheLinkedList extends App {

  def middleNode(head: ListNode): ListNode = {
    @annotation.tailrec
    def run(list1: ListNode, list2: ListNode): ListNode =
      (list1, list2) match {
        case (_, l2) if l2 == null || l2.next == null => list1
        case _ => run(list1.next, list2.next.next)
      }
    run(head, head)
  }

  println(middleNode(list0))

}

File: src/main/scala/kim/dencka/java/easy/MiddleOfTheLinkedList.java
package kim.dencka.java.easy;

import kim.dencka.java.payloads.ListNode;

class MiddleOfTheLinkedList {
    public static void main(String[] args) {
        ListNode input = ListNode.fromArray(1, 2, 3, 4, 5);
        System.out.println(middleNode(input)); // [3,4,5]
    }

    // Two Pointers
    public static ListNode middleNode(ListNode head) {
        ListNode slow = head;
        ListNode fast = head;
        while (fast.next != null) {
            if (fast.next.next == null) {
                return slow.next;
            }
            slow = slow.next;
            fast = fast.next.next;
        }
        return slow;
    }

    //Iterative
//    public static ListNode middleNode(ListNode head) {
//        ListNode list = head;
//        int depth = 0;
//        while (list != null) {
//            depth++;
//            list = list.next;
//        }
//        depth = depth / 2;
//        while (depth > 0) {
//            head = head.next;
//            depth--;
//        }
//        return head;
//    }

}